test/vfd_swmr_attrdset_writer.c

```diff
@@ -0,0 +1,2015 @@
+        if(!verify_attr_action(action, ds->fa_did, which))
+            ++nerrors;
+        if(s->csteps == 1 && (action != MODIFY_ATTR)) {
+            if(!verify_storage_cont(action, ds->fa_did, which, ds->fa_max_compact, ds->fa_min_dense, s->asteps))
+                ++nerrors;
+        }
+
+        HDassert(ds->ea_did != badhid);
+        dbgf(2, "Verifying attribute to chunked dataset: ea index\n");
+        if(!verify_attr_action(action, ds->ea_did, which))
+            ++nerrors;
+        if(s->csteps == 1 && (action != MODIFY_ATTR)) {
+            if(!verify_storage_cont(action, ds->ea_did, which, ds->ea_max_compact, ds->ea_min_dense, s->asteps))
+                ++nerrors;
+        }
+
+        HDassert(ds->bt2_did != badhid);
+        dbgf(2, "Verifying attribute to chunked dataset: bt2 index\n");
+        if(!verify_attr_action(action, ds->bt2_did, which))
+            ++nerrors;
+        if(s->csteps == 1 && (action != MODIFY_ATTR)) {
+            if(!verify_storage_cont(action, ds->bt2_did, which, ds->bt2_max_compact, ds->bt2_min_dense, s->asteps))
+                ++nerrors;
+        }
+    }
+
+    if(nerrors)
+        ret = false;
+
+    return (ret);
+
+} /* verify_attr_dsets_action() */
+
+/*
+ * Verify the attribute action on the specified dataset.
+ */
+static bool
+verify_attr_action(unsigned action, hid_t did, unsigned which)
+{
+    char name[sizeof("attr-9999999999")];
+    bool ret;
+
+    esnprintf(name, sizeof(name), "attr-%u", which);
+
+    switch(action) {
+        case ADD_ATTR: 
+            ret = verify_add_or_modify_attr(action, did, name, which);
+            break;
+
+        case MODIFY_ATTR: 
+            ret = verify_add_or_modify_attr(action, did, name, which);
+            break;
+
+        case DELETE_ATTR: 
+            ret = verify_delete_attr(did, name);
+            break;
+        
+        default:
+            HDassert(0 && "Unknown action?!?");
+    } /* end switch */
+
+    return ret;
+} /* verify_attr_action() */
+
+/* 
+ * Verify the attribute is added or modified
+ */
+static bool
+verify_add_or_modify_attr(unsigned action, hid_t did, char *attr_name, unsigned int which)
+{
+    unsigned int read_which;
+    char vl_which[sizeof("attr-9999999999")];
+    char *read_vl_which;
+    bool is_vl = false;
+    hid_t aid, atid;
+    bool ret;
+
+    HDassert(did != badhid);
+    HDassert(action == ADD_ATTR || action == MODIFY_ATTR);
+
+    if ((aid = H5Aopen(did, attr_name, H5P_DEFAULT)) < 0) {
+        printf("H5Aopen failed\n");
+        TEST_ERROR;
+    }
+
+    if((atid = H5Aget_type(aid)) < 0) {
+        printf("H5Aget_type failed\n");
+        TEST_ERROR;
+    }
+
+    if((is_vl = H5Tis_variable_str(atid))) {
+        if(action == ADD_ATTR)
+            HDsprintf(vl_which, "%u", which);
+        else
+            HDsprintf(vl_which, "%u %u %u %u %u", which, which+1, which+2, which+3, which+4);
+
+        read_vl_which = HDmalloc(sizeof("9999999999"));
```

Comment:
This allocated memory needs to be freed.

Reply:
Corrected.